### PR TITLE
Add ACPI support based on ACPICA library

### LINKED ISSUE #56919: Add ACPI bus driver support using acpica external open source project
ACPI is an hardware abstraction interface for abstract platform HW complexity with the help of platform FW (eg: BIOS) and provide a unified interface to OS so that OS can be easily ported with minimum modification to different platforms (which are ACPI complaint). 

ACPI will significantly reduce Zephyr porting effort on newer platforms since it will identify supported devices and cpu/platform/device resource and automatically configure them dynamically during system boot. Please note that most of current DTS information can be removed since acpi will provide those information on run time. The device and platform configuration for these platforms normally provided via BIOS. Hence user can go and modify them via BIOS (or even BIOS menu in some cases) without re-build of Zephyr image.

Currently Zephyr does not have a full fledge acpi support and this limit its capability to supports platforms which are acpi complaint. 

> **Note**
Please note this already discussed in Arch WG few week before as part of https://github.com/zephyrproject-rtos/zephyr/pull/56347 PR review.

The solution here proposed using an open source project namely acpica and using it as external module for realize the acpi bus driver functionality. 

Following block diagram depict overall Zephyr BSP architecture based on ACPI sub system:

Please note that "zephyr acpi bus interface" will provide a generic interface to other Zephyr sub systems/drivers and also provide some glue logics for abstract under lying acpica instantiate and complexity. The acpica is an OS-independent implementation of ACPI and hence OS Service layer will provide necessary logic to port acpica into Zephyr. The OS Service layer for Zephyr already upstreamed into acpica opensource tree. 

image

Following are some of the ACPI bus driver functionalities:
- Enumerate PCI bus/bridge and acpi device such as WDT, RTC, GPIO, APIC, DMA etc.
- Enumerate device resources such as MMIO or IO address, IRQ details
- Retrieve IO configurations such as clk freq, boud rate, polarity etc. for SOC device like I2C, SPI, UART, SMBus etc.
- Bus and other config details for on board sensors
- Retrieve platform configuration such as PCI legacy interrupt routing, APIC
- Retrieve CPU configurations/Information
- Device/platform power management and configuration

**Pull Request Details**

The https://github.com/zephyrproject-rtos/zephyr/pull/56347 PR created for enable the acpi support for Zephyr using acpica open source project.

Following commits provided:

add acpi bus driver for zephyr: https://github.com/zephyrproject-rtos/zephyr/pull/56347/commits/1040137ecdc3bcee0e44e3c5c61b36dee93064ef

build: west: add acpica module into zephyr project: https://github.com/zephyrproject-rtos/zephyr/pull/56347/commits/4bacb86267e9cfb41018374f9ecb7398b1906785

board: x86: rpl_crb: added acpi support for raptor lake crb: https://github.com/zephyrproject-rtos/zephyr/pull/56347/commits/d2dd350094ba5c01c90595ccdb9780b5f1e19b23

drivers: watchdog: add acpi support for tco wdt: https://github.com/zephyrproject-rtos/zephyr/pull/56347/commits/ae9dcdd5edd5790e232924d8a98d3c50f90f53e3

drivers: pcie: add prt support for pci legacy interrupt: https://github.com/zephyrproject-rtos/zephyr/pull/56347/commits/952885a16adb9d5c4f17572968643b98325c57dc

### PR DESCRIPTION
**Pull request by najumon1980**
Add ACPI driver support for Zephyr using acpica open source project. ACPI driver use to discover and configure platform hardware components, perform power management (e.g. putting unused hardware components to sleep), auto configuration (e.g. Plug and Play and hot swapping) etc. Though acpica support full-fledged acpi driver functionalities, current scope limited to device enumeration and retrieve acpi device resource such as MMIO address and IRQ details, PCI interrupt routing information etc. 

Fixes #56919

### CODE REVIEW

File: modules/acpi/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+	zephyr_library_sources(${COMP_DIR}/resources/rscalc.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rscreate.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsdumpinfo.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsinfo.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsio.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsirq.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rslist.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsmemory.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsmisc.c)
+	zephyr_library_sources(${COMP_DIR}/resources/rsserial.c)
```
Comment (finikorg) [inline modules/acpi/CMakeLists.txt:170]: Do you need to include only specific  files or we can use `GLOB` like:
```
file(GLOB resource_srcs ${COMP_DIR}/resources/*.c)
...
zephyr_library_sources(
     ...
    ${resource_srcs}
    ...
)
```

File: west.yml [hunk 2]
```diff
@@ -21,10 +21,18 @@ manifest:
...
+    - name: acpi
+      url-base: https://github.com/acpica
 
   #
   # Please add items below based on alphabetical order
   projects:
+    - name: acpica
+      revision: master
+      remote: acpi
+      path: modules/hal/acpi
```
[the remark this answers is not in the document]
Comment (tbursztyka) [reply west.yml:33]: Same thing here yes:
- acpi_* prefix for API
- modules/lib/acpica for the tree (indeed modules/hal is not the right place)

**Comment by carlescufi** [comment]
Architecture WG:

- @nashif states that this should not be in `include/drivers`, because it's not an actual hardware driver
- @galak agrees, maybe this should be in `subsys/` or even `modules/acpica`

**Comment by nashif** [comment]
how does the api being introduced relate to `include/zephyr/arch/x86/acpi.h`?

**Comment by najumon1980** [comment]
> Architecture WG:
> 
> * @nashif states that this should not be in `include/drivers`, because it's not an actual hardware driver
> * @galak agrees, maybe this should be in `subsys/` or even `modules/acpica`

Though we use external acpica module for realize the acpi bus driver functionality, I believe it is still a bus driver and we should treat same way as other popular OS treat it. For example FreeBSD use acpica for realize acpi bus driver functionality. Also by definition ACPI is an hardware abstraction interface for abstract platform HW complexity with the help of platform FW (eg: BIOS) and provide a unified interface to OS so that OS can be easily ported with minimum modification to different platforms (which are ACPI complaint). We can still consider this model as HAL modules where ACPICA is similar to MCU vendor specific code which abstract HW.

To understand the ACPI bus driver concept better , I can provide following details of some of the ACPI bus driver functionalities:
- Enumerate PCI bus/bridge and acpi device such as WDT, RTC, GPIO, APIC, DMA etc.
- Enumerate device resources such as MMIO or IO address, IRQ details
- Retrieve IO configurations such as clk freq, boud rate, polarity etc. for SOC device like I2C, SPI, UART, SMBus etc.
- Bus and other config details for on board sensors
- Retrieve platform configuration such as PCI legacy interrupt routing, APIC
- Retrieve CPU configurations/Information
- Device/platform power management and configuration

ACPI will significantly reduce Zephyr porting effort on newer platforms since it will identify supported devices and cpu/platform/device resource and automatically configure them dynamically during system boot. Please note that most of current DTS information can be removed since acpi will provide those information on run time. The device and platform configuration for these platforms normally provided via BIOS. Hence user can go and modify them via BIOS menu which help to avoid re-build of Zephyr image.

Following block diagram depict overall Zephyr BSP architecture based on ACPI sub system:

Please note that "zephyr acpi bus interface" will provide a generic interface to other Zephyr sub systems/drivers and also provide some glue logics for abstract under lying acpica instantiate and complexity. The acpica is an OS-independent implementation of ACPI and hence OS Service layer will provide necessary logic to port acpica into Zephyr. The OS Service layer for Zephyr already upstreamed into acpica opensource tree.

image

**Comment by najumon1980** [comment]
> how does the api being introduced relate to `include/zephyr/arch/x86/acpi.h`?

We will replace these APIs with the new "acpi bus interface" in succeeding releases (mostly as part of RPL itself or up coming Intel platforms). 
Also we have future plan for recently upstreamed drivers such as GPIO, RTC, WDT, PWM, TGPIO etc. update with new "acpi bus interface" for resource enumerations instead of hard coding via DTS. Even the PCI device like I2C, SPI, UART (clk freq, baud rate, polarity etc.), TSN GBE configuration  also can be retrieved from ACPI instead of hard coding via DTS.

**Review by tbursztyka — DISMISSED** [review]
2 major comments:

- remove any mention of "bus", from an ACPI perspective: this is just an interface. At least the specs do not specifically tell that this is a bus (let me know if I am mistaken here)
- As far as I can see, ACPI should probably not be part of DTS at all. (So even the acpi node properties I am talking in the comments, probably they can go away). The way Zephyr use DTS is built-time only (unlike it's usage in Linux where the binary embeds the device tree), where ACPI is a fully runtime-centered interface. So when CONFIG_ACPI=y, drivers should be using that info relevantly at runtime, and must be able to revert back to the built-time provided info from DTS.

File: arch/x86/Kconfig [hunk 3]
```diff
@@ -227,22 +227,22 @@ choice
 		  Reboot via the RST_CNT register, going back to BIOS.
 endchoice
 
-config ACPI
+config x86_ACPI
```
Comment (tbursztyka) [inline arch/x86/Kconfig:230]: all capital letter

**Review by tbursztyka — COMMENTED** [review]
just a few style comments, overall it looks much better now

File: drivers/Kconfig [hunk 4]
```diff
@@ -79,5 +79,7 @@ source "drivers/w1/Kconfig"
 source "drivers/watchdog/Kconfig"
 source "drivers/wifi/Kconfig"
 source "drivers/xen/Kconfig"
+source "drivers/rtc/Kconfig"
```
Comment (jhedberg) [inline drivers/Kconfig:82]: This is a duplicate of line 67, and doesn’t anyway belong in this PR

File: doc/releases/release-notes-2.5.rst
Comment (jhedberg) [inline doc/releases/release-notes-2.5.rst:1]: You shouldn’t need to modify old release notes files.

**Review by jhedberg — CHANGES_REQUESTED** [review]
A couple of small comments

File: west.yml [hunk 5]
```diff
@@ -32,6 +32,9 @@ manifest:
       revision: 908ffde6298a937c6549dbfa843a62caab26bfc5
       import:
         path-prefix: tools
+    - name: acpica
```
Comment (jhedberg) [inline west.yml:35]: It looks like all modules are in alphabetical order, so you should move this above bsim

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: drivers/acpi/acpi.c [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+	status = AcpiNsInternalizeName(name, &path);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("Invalid namestring: %s", name);
+		return NULL;
+	}
+
+	status = AcpiNsLookup(NULL, path, ACPI_TYPE_ANY, ACPI_IMODE_EXECUTE,
+			      ACPI_NS_NO_UPSEARCH | ACPI_NS_DONT_OPEN_SCOPE, NULL, &node);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("Could not locate name: %s, %d", name, status);
```
Comment (finikorg) [inline drivers/acpi/acpi.c:135]: should you return `NULL` here?

File: drivers/acpi/acpi.c [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+	status = AcpiGetCurrentResources(node, &rt_buffer);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("AcpiGetCurrentResources failed: %s", AcpiFormatException(status));
+	} else {
+		child_dev->res_lst = rt_buffer.Pointer;
+	}
+
+exit:
+
+	return 0;
```
Comment (finikorg) [inline drivers/acpi/acpi.c:276]: should you return `status` here?

File: drivers/pcie/host/pcie.c [hunk 8]
```diff
@@ -277,14 +281,25 @@ unsigned int pcie_alloc_irq(pcie_bdf_t bdf)
...
+#endif
 
 	data = pcie_conf_read(bdf, PCIE_CONF_INTR);
 	irq = PCIE_CONF_INTR_IRQ(data);
 
 	if (irq == PCIE_CONF_INTR_IRQ_NONE ||
 	    irq >= CONFIG_MAX_IRQ_LINES ||
 	    arch_irq_is_used(irq)) {
+#ifdef CONFIG_ACPI
+		dev = device_get_binding("acpi");
```
Comment (jhedberg) [inline drivers/pcie/host/pcie.c:295]: Just move the `const struct device *dev` here. That way you don't need to have two #ifdef blocks in this function.

File: drivers/acpi/acpi.c [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+	int status;
+	struct acpi *bus = dev->data;
+
+	LOG_DBG("");
+
+	/* For debug version only */
+	ACPI_DEBUG_INITIALIZE();
+
+	status = initialize_acpica();
+	if (status != AE_OK) {
```
Comment (finikorg) [inline drivers/acpi/acpi.c:300]: Why to use different methods for error check? `status != AE_OK` vs `ACPI_FAILURE(status)`? Please use one method in the same code. Lets use `ACPI_FAILURE()` and `ACPI_SUCCESS()`

File: drivers/acpi/acpi.c [hunk 10]
```diff
@@ -0,0 +1,543 @@
...
+		LOG_WRN("Error in enable pic mode acpi method:%d", status);
+	}
+
+	status = acpi_retrieve_legacy_irq(bus);
+	if (status) {
+		LOG_ERR("Error in retrieve legacy interrupt info:%d", status);
+		goto exit;
+	}
+	bus->status = AE_OK;
```
Comment (finikorg) [inline drivers/acpi/acpi.c:316]: minor: can be `bus->status = status` (and `return status` later)

**Review by andyross — COMMENTED** [review]
Some notes.  The big one is the design point about interrupts: unless I've misunderstood how acpica is supposed to work, we can't use the ACPI interrupt facilities without major and invasive changes to how x86 interrupt handling works in Zephyr.  I don't believe Linux does that either.

ACPI is limited to data traversal (it's main role) and very occasionally simple MMIO in evaluated nodes.

File: drivers/acpi/acpi.c [hunk 11]
```diff
@@ -0,0 +1,543 @@
...
+	}
+
+	if (handle->Type != ACPI_TYPE_DEVICE) {
+		LOG_ERR("No ACPI node foud with given name: %s", bus_name);
+		goto exit;
+	}
+
+	node = ACPI_CAST_PTR(ACPI_NAMESPACE_NODE, handle);
+
+	(void)AcpiGetHandle(node, method, ACPI_CAST_PTR(ACPI_HANDLE, &prt_node));
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/acpi/acpi.c:163]: I don't think it is really needed to call AcpiEvaluateObject() for AcpiGetIrqRoutingTable(), AcpiRsGetCrsMethodData() etc. since it already invoke internal function AcpiUtEvaluateObject ().  Anyway I will check with ACPICA team on this.

File: drivers/acpi/acpi.c [hunk 12]
```diff
@@ -0,0 +1,543 @@
...
+	if (!prt_node) {
+		LOG_ERR("No entry for the ACPI node with given name: %s", bus_name);
+		goto exit;
+	}
+	return node;
+exit:
+	return NULL;
+}
+
+static ACPI_STATUS acpi_enable_pic_mode(void)
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/acpi/acpi.c:174]: This is only to inform the ASL code which interrupt controller is being used

File: drivers/acpi/acpi.c [hunk 13]
```diff
@@ -0,0 +1,543 @@
...
+	node = acpi_evaluate_method(bus_name, METHOD_NAME__PRT);
+	if (!node) {
+		LOG_ERR("Evaluation failed for given device: %s", bus_name);
+		return -ENODEV;
+	}
+
+	rt_buffer.Pointer = rt_table;
+	rt_buffer.Length = ACPI_DEBUG_BUFFER_SIZE;
+
+	status = AcpiGetIrqRoutingTable(node, &rt_buffer);
```
[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/acpi/acpi.c:211]: This is for retrieve PCH interrupt routing information for PCI devices using _PRT method. We will check IRQ information provided in PCI config space and if not present, then try to retrieve via ACPI and configure PCI interrupt ctrl register. Currently if CONFIG_ACPI enabled then it will not invoke arch_irq_allocate() (please see pcie.c). But looks we need to mark IRQ numbers retrieved via PRT as allocated/reserved to avoid conflict if someone try to allocated arch_irq_allocate() such as for MSI. I will try to figure out this and update the PR.

[the remark this answers is not in the document]
Comment (najumon1980) [reply drivers/acpi/acpi.c:213]: updated code to mark PRT IRQs as reserved using arch_irq_set_used (please see line number 219).

File: drivers/acpi/acpi.c [hunk 14]
```diff
@@ -0,0 +1,554 @@
...
+	status = AcpiInitializeObjects(ACPI_FULL_INITIALIZATION);
+	if (ACPI_FAILURE(status)) {
+		ACPI_EXCEPTION((AE_INFO, status, "While initializing ACPI objects"));
+	}
+exit:
+
+	return status;
+}
+
+static ACPI_NAMESPACE_NODE *acpi_name_Lookup(char *name)
```
Comment (jhedberg) [inline drivers/acpi/acpi.c:118]: Lower-case l

File: drivers/acpi/acpi_shell.c [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+{
+	struct acpi_dev *dev;
+
+	if (argc < 2) {
+		return -EINVAL;
+	}
+
+	acpi_bus = device_get_binding("acpi");
+	if (!acpi_bus) {
+		printk("No ACPI driver found\n");
```
Comment (jhedberg) [inline drivers/acpi/acpi_shell.c:255]: Please use shell output functions instead of printk

**Comment by najumon1980** [comment]
@andyross, could please review my response and also the updated code ?

File: dts/bindings/base/base.yaml [hunk 16]
```diff
@@ -19,6 +19,18 @@ properties:
     required: true
     description: compatible strings
 
+  hardware-id:
```
Comment (finikorg) [inline dts/bindings/base/base.yaml:22]: How do you use it?

File: drivers/acpi/acpi_shell.c [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+
+	return 0;
+}
+
+static int dis_dbg_prints(const struct shell *sh, size_t argc, char **argv)
+{
+	shell_fprintf(sh, SHELL_NORMAL, "Disable Debug logs\n");
+
+	return 0;
+}
```
Comment (jhedberg) [inline drivers/acpi/acpi_shell.c:215]: These don't seem to be doing anything useful, so just remove them.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: dts/bindings/base/base.yaml (hunk 16 above)
Comment (finikorg) [reply dts/bindings/base/base.yaml:22]: Is this change related to your other PR?

File: drivers/acpi/acpi.c [hunk 18]
```diff
@@ -0,0 +1,554 @@
...
+
+	return index < bus->num_dev ? &bus->child_dev[index] : NULL;
+}
+
+static struct acpi bus_ctx = {
+	.status = AE_NOT_CONFIGURED,
+};
+
+/* TODO: need to enable initialization levels as PRE_KERNEL_1. */
+DEVICE_DEFINE(acpi, "acpi", acpi_init, NULL, &bus_ctx, NULL, POST_KERNEL,
```
Comment (tbursztyka) [inline drivers/acpi/acpi.c:553]: Actually, acpi is not a device per se. 

So, either you call SYS_INIT() or, better because this would not end up dependent on a level/prio, make acpi_init() public and let the arch to call it at the right place (currently, only x86 would). (perhaps that would mean also acpi should not be in drivers/ but in lib/ or something like that)

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/acpi/acpi.c (hunk 18 above)
Comment (najumon1980) [reply drivers/acpi/acpi.c:553]: SYS_INIT is also have level/prio and order of invoked based on it.

File: dts/bindings/base/base.yaml (hunk 16 above)
Comment (najumon1980) [reply dts/bindings/base/base.yaml:22]: This is required for ACPI since acpi devices will be enumerated based on it.

File: drivers/acpi/acpi.c (hunk 18 above)
Comment (najumon1980) [reply drivers/acpi/acpi.c]: updated with SYS_INIT

Comment (tbursztyka) [reply drivers/acpi/acpi.c]: > SYS_INIT is also have level/prio and order of invoked based on it.

Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

Comment (jhedberg) [reply drivers/acpi/acpi.c]: > Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

This sounds like a good approach for me as well, i.e. do the init from the arch code. Btw, is `drivers/acpi/` the right location? Or would `lib/acpi/` or `subsys/acpi/` make more sense?

File: drivers/acpi/acpi.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (andyross) [reply drivers/acpi/acpi.c:168]: Then why is this function called "acpi_evaluate_method" if it's not evaluating anything?  Name it "acpi_get_bus_method" or somesuch?

File: drivers/acpi/acpi.c (hunk 18 above)
Comment (najumon1980) [reply drivers/acpi/acpi.c]: ACPI is mostly deal with platform HW and devices. So my view is that it should be part of drivers since it mostly use by drivers. For other OS like Linux and Windows is exposed as drivers.

Comment (najumon1980) [reply drivers/acpi/acpi.c]: > > SYS_INIT is also have level/prio and order of invoked based on it.
> 
> Thus the better proposal, on the same comment: making acpi_init() public and calling it directly in arch/x86/ relevant place.

For CPU configurations etc. acpica provided initialization sequence for early ACPI table access etc. before doing full acpica initialization. This can be done as part of x86 arch porting update for using acpica APIs and currently it is enabled only for device drivers.

Comment (tbursztyka) [reply drivers/acpi/acpi.c]: > ACPI is mostly deal with platform HW and devices. So my view is that it should be part of drivers since it mostly use by drivers. For other OS like Linux and Windows is exposed as drivers.

It is _used_ by drivers but acpi itself is not a device and thus the acpica is more like a lib or a subsys.

Comment (najumon1980) [reply drivers/acpi/acpi.c]: moved acpi under lib/acpi.

File: include/zephyr/acpi/acpi.h [hunk 19]
```diff
@@ -0,0 +1,130 @@
+/*
```
Comment (tbursztyka) [inline include/zephyr/acpi/acpi.h:1]: Change the commit title "lib: Add acpica library" or something like that. Change the message accordingly (i.e. remove "driver" word"

File: drivers/serial/Kconfig.ns16550 [hunk 20]
```diff
@@ -72,6 +72,11 @@ config UART_NS16550_SIMULT_ACCESS
 	  When enabled, NS16550 supports IO, MMIO, PCIe UART devices simultaneously.
 	  For io-mapped instances, io-mapped DTS property need to be added in dtsi.
 
+config UART_NS16550_AUTO_BOOT_LEVEL
```
Comment (tbursztyka) [inline drivers/serial/Kconfig.ns16550:75]: This commit seems unrelated to ACPI, move it to another PR

**Review by tbursztyka — COMMENTED** [review]
small comments,  otherwise lgtm

**Review by tbursztyka — DISMISSED** [review]
lgtm

File: include/zephyr/acpi/acpi.h [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+ * Copyright (c) 2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_DRIVERS_ACPI_H_
+#define ZEPHYR_INCLUDE_DRIVERS_ACPI_H_
+#include <acpi.h>
```
Comment (jhedberg) [inline include/zephyr/acpi/acpi.h:8]: Having Zephyr `acpi.h` include another `acpi.h` is pretty confusing. There might also be the risk for infinite recursion if the referenced file ends up being this same file. I believe the reference is to the `acpi.h` header file in the ACPICA tree, correct? Would it be possible to set the include lookup paths in CMakeLists.txt in such a way that you need to provide some name pacing for the include statement here, e.g. `#include <acpica/acpi.h>`?

**Comment by najumon1980** [comment]
Hi All, 
kindly request to approve the PR if no more comments and is pending for a while now. Also this PR currently blocking other IO drivers and it will be great if you can expedite the review.

File: lib/acpi/acpi.c [hunk 22]
```diff
@@ -0,0 +1,623 @@
...
+
+	acpi_enum_devices(&bus_ctx);
+
+exit:
+	bus_ctx.status = status;
+
+	return status;
+}
+
+int acpi_early_init(void)
```
Comment (jhedberg) [inline lib/acpi/acpi.c:370]: Where is this function called from? I don't see any use of it in this PR.

File: modules/acpica/CMakeLists.txt [hunk 23]
```diff
@@ -0,0 +1,177 @@
...
+set(ACPI_DIR ${ZEPHYR_CURRENT_MODULE_DIR}/)
+set(INC_DIR ${ACPI_DIR}/source/include/)
+set(SRC_DIR ${ACPI_DIR}/source)
+set(COMP_DIR ${ACPI_DIR}/source/components)
+set(PARENT_SRC_DIR ${ACPI_DIR}../../zephyr)
+set(ACPI_PARENT_DIR ${ACPI_DIR}/../)
+
+zephyr_include_directories(
+	${PARENT_SRC_DIR}/include/
+	${ACPI_PARENT_DIR}/
```
Comment (jhedberg) [inline modules/acpica/CMakeLists.txt:13]: This is the same as `modules/lib` in the west project directory hierarchy, right? I do think it's an improvement that you have some directory namespacing for the ACPICA headers, but I'm not sure it's ideal that we add a new include lookup path for a directory that's this generic (not specific to acpica). That said, it might be good enough for an initial solution, so we get this merged and can improve the situation in subsequent PRs.

File: lib/acpi/acpi.c (hunk 22 above)
Comment (najumon1980) [reply lib/acpi/acpi.c:370]: This API is useful for early platform init such as for CPU and currently it is given as placeholder

Comment (jhedberg) [reply lib/acpi/acpi.c:370]: Understood, however I think it would be better to add it only once you have a place that will call it, i.e. in some later PR. This will help explain its purpose as well as ensure proper code/testing coverage.

Comment (najumon1980) [reply lib/acpi/acpi.c:370]: Add shell command for retrieve acpi table and this will invoke acpi_early_init.

File: include/zephyr/acpi/acpi.h [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * @brief retrieve acpi table for the given signature.
+ *
+ * @param Signature pointer to the 4-character ACPI signature for the requested table
+ * @param inst instance number for the requested table
+ * @param acpi_table pointer to the acpi table
+ * @return return 0 on success or error code
+ */
+int acpi_table_get(char *Signature, int inst, void **acpi_table);
```
Comment (jhedberg) [inline include/zephyr/acpi/acpi.h:123]: Why capital `S` in `Signature`?

Comment (jhedberg) [reply include/zephyr/acpi/acpi.h:123]: Btw, shouldn't it be `const char *`?

File: lib/acpi/acpi.c (hunk 22 above)
Comment (jhedberg) [reply lib/acpi/acpi.c:370]: The shell runs very late in the boot, and since you're entering commands it's pretty much guaranteed that everything happens *after* the boot. So I'm a bit confused by something called "early init" would be appropriate to be called from there.

Comment (najumon1980) [reply lib/acpi/acpi.c:370]: acpi_early_init  verified locally in pre-kernel stage locally.

Comment (finikorg) [reply lib/acpi/acpi.c:370]: > acpi_early_init verified locally in pre-kernel stage locally.

Cannot understand this sentence.

File: include/zephyr/acpi/acpi.h [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+ *
+ * @param res the list of acpi resource list
+ * @return resource type on success or invalid resource type
+ */
+int acpi_device_type_get(ACPI_RESOURCE *res);
+
+/**
+ * @brief retrieve acpi table for the given signature.
+ *
+ * @param Signature pointer to the 4-character ACPI signature for the requested table
```
Comment (jhedberg) [inline include/zephyr/acpi/acpi.h:118]: Still `Signature` here (this will generate documentation generation errors)

**Comment by najumon1980** [comment]
Update ACPICA module to latest code base

**Review by jhedberg — APPROVED** [review]
This looks good enough to me now to be merged as an initial implementation

**Review by finikorg — APPROVED** [review]

**Review by tbursztyka — APPROVED** [review]
